test/e2e/tcpudp/tcp.go

```diff
@@ -86,12 +94,104 @@ var _ = framework.IngressNginxDescribe("TCP Feature", func() {
 
 		ip := f.GetNginxIP()
 		port, err := f.GetNginxPort("http-svc")
-		Expect(err).NotTo(HaveOccurred(), "unexpected error obtaning service port")
+		Expect(err).NotTo(HaveOccurred(), "unexpected error obtaining service port")
 
 		resp, _, errs := gorequest.New().
 			Get(fmt.Sprintf("http://%v:%v", ip, port)).
 			End()
 		Expect(errs).Should(BeEmpty())
 		Expect(resp.StatusCode).Should(Equal(200))
 	})
+
+	It("should expose an ExternalName TCP service", func() {
+		// Setup:
+		// - Create an external name service for DNS lookups on port 5353. Point it to google's DNS server
+		// - Expose port 5353 on the nginx ingress NodePort service to open a hole for this test
+		// - Update the `tcp-services` configmap to proxy traffic to the configured external name service
+
+		// Create an external service for DNS
+		externalService := &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "dns-external-name-svc",
+				Namespace: f.IngressController.Namespace,
+			},
+
+			Spec: corev1.ServiceSpec{
+				ExternalName: "google-public-dns-a.google.com",
+				Ports: []corev1.ServicePort{
+					{
+						Name:       "dns-external-name-svc",
+						Port:       5353,
+						TargetPort: intstr.FromInt(53),
+					},
+				},
+				Type: corev1.ServiceTypeExternalName,
+			},
+		}
+		f.EnsureService(externalService)
+
+		// Expose the `external name` port on the `ingress-nginx` service
+		svc, err := f.KubeClientSet.
+			CoreV1().
+			Services(f.IngressController.Namespace).
+			Get("ingress-nginx", metav1.GetOptions{})
+		Expect(err).To(BeNil(), "unexpected error obtaining ingress-nginx service")
+		Expect(svc).NotTo(BeNil(), "expected a service but none returned")
+
+		svc.Spec.Ports = append(svc.Spec.Ports, corev1.ServicePort{
+			Name:       "dns-svc",
+			Port:       5353,
+			TargetPort: intstr.FromInt(5353),
+		})
+		_, err = f.KubeClientSet.
+			CoreV1().
+			Services(f.IngressController.Namespace).
+			Update(svc)
+		Expect(err).NotTo(HaveOccurred(), "unexpected error updating service")
+
+		// Update the TCP configmap to link port 5353 to the DNS external name service
```

Comment:
Can you use `UpdateNginxConfigMapData` helper function instead?

Reply:
I took a look at `UpdateNginxConfigMapData`, however it appears hard coded to update the `nginx-configuration` configmap (copied below).  This bit of test code here needs to update the `tcp-services` configmap.  Would you suggest extending `UpdateNginxConfigMapData` to accept a parameter with the configmap name or leave as is?

```
config, err := f.KubeClientSet.
		CoreV1().
		ConfigMaps(f.IngressController.Namespace).
		Get("nginx-configuration", metav1.GetOptions{})
```

Reply:
You are right, I missed that.